oak_functions/loader/src/server.rs

```diff
@@ -87,69 +86,58 @@ impl Policy {
-
-        if required_padding < 0 {
-            anyhow::bail!("constant_response_size_bytes specified by the policy is too small.")
-        } else {
-            let padding_bytes = vec![b'0'; required_padding as usize];
-            http::response::Builder::new()
-                .status(StatusCode::INTERNAL_SERVER_ERROR)
-                .header(POLICY_PADDING_HEADER, std::str::from_utf8(&padding_bytes)?)
-                .body(body.into())
-                .context("couldn't create response")
-        }
-    }
-
-    /// Pads the response by adding a number of unspecified characters, to make the total size of
-    /// the response equal to `self.constant_response_size_bytes`
+    /// Creates an HTTP response, with status `OK`, the input `response` as its body.
     ///
-    /// The zeros are added as the header value for header `oak-policy-padding`.
-    /// Returns an error if the size of the input `response` is larger than or equal to
-    /// `self.constant_response_size_bytes`.
-    fn pad_response(&self, response: FunctionsResponse) -> anyhow::Result<Response<Body>> {
-        let response_size = estimate_size(&response)?;
-        let required_padding = self.constant_response_size_bytes as i64 - response_size;
-        if required_padding < 0 {
-            panic!(
-                "Something is very wrong. Only call `pad_response` with responses that require padding."
-            )
+    /// To conform to the security policy, the response is padded by adding a number of unspecified
+    /// characters, to make the total size of the response equal to
+    /// `self.constant_response_size_bytes`. The padding is added as the value of the
+    /// `oak-policy-padding` header. If the size of the input `response` is larger than
+    /// `self.constant_response_size_bytes` and the response is not already indicating policy
+    /// violation, then the `response` is discarded and an HTTP response indicating the policy
+    /// violation is returned. If `self.constant_response_size_bytes` is too small even for the
+    /// policy violation response, or the response already indicates a policy violation, an error is
+    /// returned.
+    fn create_http_response(&self, response: FunctionsResponse) -> anyhow::Result<Response<Body>> {
+        // deserialize response and get its size.
+        let mut encoded = vec![];
+        response
+            .encode(&mut encoded)
+            .context("couldn't encode response")?;
+        let padding_length = self.constant_response_size_bytes as i64 - encoded.len() as i64;
+        if padding_length < 0 {
+            match FunctionsStatusCode::from_i32(response.status) {
+                None => anyhow::bail!("invalid status code"),
+                Some(FunctionsStatusCode::PolicyViolation) => anyhow::bail!(
+                    "constant_response_size_bytes specified by the policy is too small."
+                ),
+                _ => self.create_http_response(FunctionsResponse {
```

Comment:
Why is this even necessary? If the response is too large (i.e. padding < 0) then just always return the policy violation message, no?

edit: I just realised that this may be because of the recursive invocation? If so, it should definitely have a clear comment explaining the recursion, this logic, and convincing a reader that it does terminate. Alternatively, I think it would be better to avoid the recursion altogether if possible.